espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt

```diff
@@ -0,0 +1,303 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.espressoserver.lib.model
+
+import android.content.Context
+import android.os.SystemClock
+import android.text.TextUtils
+import android.util.SparseArray
+import android.util.Xml
+import android.view.View
+import android.view.ViewGroup
+import android.widget.AdapterView
+import androidx.test.core.app.ApplicationProvider.getApplicationContext
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
+import io.appium.espressoserver.lib.handlers.exceptions.XPathLookupException
+import io.appium.espressoserver.lib.helpers.AndroidLogger.logger
+import io.appium.espressoserver.lib.helpers.StringHelpers.abbreviate
+import io.appium.espressoserver.lib.helpers.XMLHelpers.toNodeName
+import io.appium.espressoserver.lib.helpers.XMLHelpers.toSafeString
+import io.appium.espressoserver.lib.viewaction.ViewGetter
+import org.w3c.dom.Element
+import org.w3c.dom.NodeList
+import org.xml.sax.InputSource
+import org.xmlpull.v1.XmlSerializer
+import java.io.*
+import java.util.*
+import java.util.concurrent.Semaphore
+import javax.xml.xpath.XPathConstants
+import javax.xml.xpath.XPathExpressionException
+import javax.xml.xpath.XPathFactory
+
+class SourceDocument @JvmOverloads constructor(
+    val root: View?,
+    val viewMap: SparseArray<View> = SparseArray()
+) {
+    private val RESOURCES_GUARD = Semaphore(1)
+
+    private var serializer: XmlSerializer? = null
+    private var tmpXmlName: String? = null
+
+    constructor() : this(null)
+
+    @Throws(IOException::class)
+    private fun setAttribute(attrName: ViewAttributesEnum, attrValue: Any?) {
+        // Do not write attributes, whose values equal to null
+        attrValue?.let {
+            // Cut off longer strings to avoid OOM errors
+            val xmlValue = abbreviate(toSafeString(it.toString(), NON_XML_CHAR_REPLACEMENT),
+                    MAX_XML_VALUE_LENGTH)
+            serializer?.attribute(NAMESPACE, attrName.toString(), xmlValue)
+        }
+    }
+
+    @Throws(IOException::class)
+    private fun recordAdapterViewInfo(adapterView: AdapterView<*>) {
+        val adapter = adapterView.adapter ?: return
+        val adapterCount = adapter.count
+        val adapterData = ArrayList<String>()
```

Comment:
use .map

Reply:
This is a different case. Not every value gets added to the `adapterData` array.

Reply:
I still think the current code could be optimised, but it's not a must